src/backend/query_builder.rs

```diff
@@ -1378,29 +1343,49 @@ pub trait QueryBuilder: QuotedBuilder + EscapeBuilder + TableRefBuilder {
         right: &SimpleExpr,
         sql: &mut dyn SqlWriter,
     ) {
-        let no_paren = matches!(op, BinOper::Equal | BinOper::NotEqual);
-        let left_paren = left.need_parentheses()
-            && left.is_binary()
-            && op != left.get_bin_oper().unwrap()
-            && !no_paren;
+        // If left has higher precedence than op, we can drop parentheses around left
+        let drop_left_higher_precedence =
+            self.inner_expr_well_known_greater_precedence(left, &(*op).into());
+
+        //Figure out if left associativity rules allow us to drop the left parenthesis
+        let drop_left_assoc = left.is_binary()
+            && op == left.get_bin_oper().unwrap()
+            && self.well_known_left_associative(op);
+
+        let left_paren = !drop_left_higher_precedence && !drop_left_assoc;
         if left_paren {
             write!(sql, "(").unwrap();
         }
         self.prepare_simple_expr(left, sql);
         if left_paren {
             write!(sql, ")").unwrap();
         }
+
         write!(sql, " ").unwrap();
         self.prepare_bin_oper(op, sql);
         write!(sql, " ").unwrap();
-        let no_right_paren = matches!(
-            op,
-            BinOper::Between | BinOper::NotBetween | BinOper::Like | BinOper::NotLike
-        );
-        let right_paren = (right.need_parentheses()
-            || right.is_binary() && op != left.get_bin_oper().unwrap())
-            && !no_right_paren
-            && !no_paren;
+
+        // If right has higher precedence than op, we can drop parentheses around right
+        let drop_right_higher_precedence =
+            self.inner_expr_well_known_greater_precedence(right, &(*op).into());
+
+        // Due to representation of trinary op between as nested binary ops.
+        let drop_right_between_hack = (op == &BinOper::Between || op == &BinOper::NotBetween)
+            && right.is_binary()
+            && right.get_bin_oper().unwrap() == &BinOper::And;
```

Comment:
Could we rewrite that as something like
matches!(right, Oper(BinOper::And))
To get rid of the unwrap

Same below

Reply:
Changed to: 
&& matches!(right.get_bin_oper(), Some(&BinOper::And)); 
Also below with escape.